### ISSUE TITLE
if statements not cleanly short-circuiting on Apple Silicon with optimizations

### ISSUE THREAD

**Issue by mckib2** [llvm/llvm-project#60695]
- x-ref https://github.com/boostorg/math/pull/945

Overflows are reported for `-O2` and `-O3` due to evaluation of unsafe calculations in if-statements that should have been guarded by a previous condition short-circuiting.  We only observe this on certain versions of MacOS, e.g., Monterrey 12.6.

Reproducer:
```cpp
// main.cpp
#include <iostream>
#include <limits>
#include <cfenv>

int main(int argc, char** argv) {
    if (argc != 2) {
        return 1;
    }
    const double x = std::atof(argv[1]);
    double f1 = std::numeric_limits<double>::max();
    double res;
    if((x < 1) && (std::numeric_limits<double>::max() * x < f1)) {
        std::cout << "if branch\n";
    } else {
        std::cout << "else branch\n";
    }

    if(std::fetestexcept(FE_OVERFLOW)) {
        std::cout << "overflow reported\n";
    } else {
        std::cout << "overflow not reported\n";
    }

    return 0;
}

// output:
//     else branch
//     overflow reported
```

Build:
```bash
clang++ -g -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk -O2 -std=gnu++14 main.cpp
```

Run:
```bash
./a.out 42.0
```

Version information:
```console
% clang++ --version
Apple clang version 14.0.0 (clang-1400.0.29.102)
Target: arm64-apple-darwin21.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Comment by fhahn** [comment]
I am unable to reproduce this with upstream clang. To report issue's with Apple Clang, please use https://feedbackassistant.apple.com/.

I am going to close this issue for now. If this reproduces with upstream Clang, please share updated instructions and I'll re-open the issue.

**Comment by mckib2** [comment]
I wouldn't have expected to see it with upstream -- this has been notoriously hard for us to get reproducible environments ourselves.  I will try submitting to apple, thanks!
